File: cucumber-expressions/javascript/lib/transform_lookup.js
const Transform = require('./transform')

class TransformLookup {
  constructor() {
    this._transformsByTypeName = new Map()
    this._transformsByCaptureGroupRegexp = new Map()
    this._transformsByConstructorName = new Map()

    const FIXNUM_REGEXPS = ["-?\\d+", "\\d+"]
    const FLOATING_POINT_REGEXPS = ["-?\\d*\\.?\\d+"]

    this.addTransform(new Transform('int', Number, FIXNUM_REGEXPS, parseInt))
    this.addTransform(new Transform('float', Number, FLOATING_POINT_REGEXPS, parseFloat))
  }

  lookupByType(type) {
    if (typeof type === 'function') {
      return this.lookupByFunction(type)
    } else if (typeof type === 'string') {
      return this.lookupByTypeName(type, false)
    } else {
      throw new Error(`Type must be string or function, but was ${type} of type ${typeof type}`)
    }
  }

  lookupByFunction(fn) {
    if (fn.name) {
      const prefix = fn.name[0]
      const looksLikeConstructor = prefix.toUpperCase() === prefix

      let transform
      if (looksLikeConstructor) {
        transform = this._transformsByConstructorName.get(fn.name)
      }
      if (!transform) {
        const factory = s => {
          if (looksLikeConstructor) {
            return new fn(s)
          } else {
            return fn(s)
          }
        }
        return this.createAnonymousLookup(factory)
      } else {
        return transform
      }
    } else {
      return this.createAnonymousLookup(fn)
    }
  }

  lookupByTypeName(s, ignoreUnknownTypeName) {
    const transform = this._transformsByTypeName.get(s)
    if (!transform) {
      if (ignoreUnknownTypeName) {
        return null
      } else {
        throw new Error(`No transformer for type name "${s}"`)
      }
    } else {
      return transform
    }
  }

  lookupByCaptureGroupRegexp(captureGroupRegexp) {
    return this._transformsByCaptureGroupRegexp.get(captureGroupRegexp)
  }

  createAnonymousLookup(fn) {
    return new Transform(null, null, [".+"], fn)
  }

  addTransform(transform) {
    this._transformsByConstructorName.set(transform.constructorFunction.name, transform)

    this._transformsByTypeName.set(transform.typeName, transform)

    transform.captureGroupRegexps.forEach(captureGroupRegexp => {
      this._transformsByCaptureGroupRegexp.set(captureGroupRegexp, transform)
    })
  }
}

module.exports = TransformLookup

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/transform_lookup.rb
require 'cucumber/cucumber_expressions/transform'

module Cucumber
  module CucumberExpressions
    class TransformLookup
      FIXNUM_REGEXPS = ['-?\d+', '\d+']
      FLOATING_POINT_REGEXPS = ['-?\d*\.?\d+']

      def initialize
        @transforms_by_type_name = {}
        @transforms_by_capture_group_regexp = {}
        @transforms_by_class = {}

        add_transform(Transform.new('int', Fixnum, FIXNUM_REGEXPS, lambda {|s| s.to_i}))
        add_transform(Transform.new('float', Float, FLOATING_POINT_REGEXPS, lambda {|s| s.to_f}))
      end

      def lookup_by_type(type)
        if type.is_a?(Class)
          lookup_by_class(type)
        elsif type.is_a?(String)
          lookup_by_type_name(type, false)
        else
          raise Exception.new("Type must be string or class, but was #{type} of type #{type.class}")
        end
      end

      def lookup_by_class(clazz)
        transform = @transforms_by_class[clazz]
        if transform.nil?
          create_anonymous_lookup(lambda {|s| clazz.new(s)})
        else
          transform
        end
      end

      def lookup_by_type_name(type_name, ignore_unknown_type_name)
        transform = @transforms_by_type_name[type_name]
        if transform.nil?
          return nil if ignore_unknown_type_name
          raise Exception.new("No transformer for type name \"#{type_name}\"")
        else
          transform
        end
      end

      def lookup_by_capture_group_regexp(capture_group_regexp)
        @transforms_by_capture_group_regexp[capture_group_regexp]
      end

      def create_anonymous_lookup(proc)
        Transform.new(nil, nil, ['.+'], proc)
      end

      def add_transform(transform)
        @transforms_by_type_name[transform.type_name] = transform
        @transforms_by_class[transform.type] = transform

        transform.capture_group_regexps.each do |capture_group_regexp|
          @transforms_by_capture_group_regexp[capture_group_regexp] = transform
        end
      end

    end
  end
end
